src/serverlist.cpp

```diff
@@ -490,6 +542,107 @@ void CServerListManager::CentralServerQueryServerList ( const CHostAddress& Inet
+    // per definition is the directory server
+    // (i.e., this server).
+    for ( int iIdx = ServerList.size() - 1; iIdx > 0; iIdx-- )
+    {
+        if ( ServerList[iIdx].HostAddr == haSearchTerm )
+        {
+            return iIdx;
+        }
+    }
+    return INVALID_INDEX;
+}
+
+void CServerListManager::CentralServerLoadServerList ( const QString strServerList )
+{
+    QFile file ( strServerList );
+
+    if ( !file.open ( QIODevice::ReadWrite | QIODevice::Text ) )
+    {
+        qWarning() << qUtf8Printable (
+            QString ( "Could not open '%1' for read/write.  Please check that Jamulus has permission (and that there is free space)." )
+                .arg ( ServerListFileName ) );
+        return;
+    }
+
+    // use entire file content for the persistent server list
+    // and remember it for later writing
+    qInfo() << "Setting persistent server list file:" << strServerList;
+    ServerListFileName = strServerList;
+    CHostAddress haServerHostAddr;
+
+    QTextStream in ( &file );
+    while ( !in.atEnd() )
+    {
+        QString     line   = in.readLine();
+        QStringList slLine = line.split ( ";" );
+        if ( slLine.count() != 7 )
+        {
+            qWarning() << qUtf8Printable ( QString ( "Could not parse '%1' successfully - bad line" ).arg ( line ) );
+            continue;
+        }
+
+        NetworkUtil::ParseNetworkAddress ( slLine[0], haServerHostAddr );
+        int iIdx = IndexOf ( haServerHostAddr );
+        if ( iIdx != INVALID_INDEX )
+        {
+            qWarning() << qUtf8Printable ( QString ( "Skipping '%1' - duplicate host %2" ).arg ( line ).arg ( haServerHostAddr.toString() ) );
+            continue;
+        }
+
+        CServerListEntry serverListEntry =
+            CServerListEntry::parse ( slLine[0], slLine[1], slLine[2], slLine[3], slLine[4], slLine[5], slLine[6].toInt() != 0 );
+
+        // We expect servers to have addresses...
+        if ( ( CHostAddress() == serverListEntry.HostAddr ) )
+        {
+            qWarning() << qUtf8Printable ( QString ( "Could not parse '%1' successfully - invalid host" ).arg ( line ) );
+            continue;
+        }
+
+        qInfo() << qUtf8Printable ( QString ( "Loading registration for %1 (%2): %3" )
+                                        .arg ( serverListEntry.HostAddr.toString() )
+                                        .arg ( serverListEntry.LHostAddr.toString() )
+                                        .arg ( serverListEntry.strName ) );
+        ServerList.append ( serverListEntry );
+    }
+}
+
+void CServerListManager::CentralServerSaveServerList()
+{
+    if ( ServerListFileName.isEmpty() )
+    {
+        return;
+    }
+
+    QFile file ( ServerListFileName );
+
+    if ( !file.open ( QIODevice::WriteOnly | QIODevice::Truncate | QIODevice::Text ) )
+    {
+        // Not a useable file
+        qWarning() << QString ( tr ( "Could not write to '%1'" ) ).arg ( ServerListFileName );
```

Comment:
Suggestion as above: ... Please check that Jamulus has permission (and that there is free space).

Reply:
No, too annoying